src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs

```diff
@@ -105,15 +114,56 @@ public class HubEndPoint<THub> : IInvocationBinder where THub : Hub
                 }
             }
 
+            _lastSendTimestamp = Stopwatch.GetTimestamp();
             var writingOutputTask = WriteToTransport();
 
+            Timer keepAliveTimer = null;
```

Comment:
We're going to have wayyyyy too many timers. We need a connection heartbeat design here. The only thing that has a connection list is the lifetime manager though. We need to think a bit more here.

Reply:
I thought we were going to look at a single timer still model.

Reply:
So did I until I realized how HubEndpoint works when reviewing it this morning :). We need to rethink this quite a bit. I think we need to go back to the slightly gross idea of having the connection manager tick the heartbeat via a feature. The problem there is the config gets gross. The ping rate is always going to be off by at most the tick rate and that gets weird.